README.md
<!------------------------------------------------------------------------------
--
-- Copyright (C) 2022 Kevin Matthes
--
-- This program is free software; you can redistribute it and/or modify
-- it under the terms of the GNU General Public License as published by
-- the Free Software Foundation; either version 2 of the License, or
-- (at your option) any later version.
--
-- This program is distributed in the hope that it will be useful,
-- but WITHOUT ANY WARRANTY; without even the implied warranty of
-- MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
-- GNU General Public License for more details.
--
-- You should have received a copy of the GNU General Public License along
-- with this program; if not, write to the Free Software Foundation, Inc.,
-- 51 Franklin Street, Fifth Floor, Boston, MA 02110-1301 USA.
--
----
--
--  FILE
--      README.md
--
--  BRIEF
--      Important information regarding this project.
--
--  AUTHOR
--      Kevin Matthes
--
--  COPYRIGHT
--      (C) 2022 Kevin Matthes.
--      This file is licensed GPL 2 as of June 1991.
--
--  DATE
--      2022
--
--  NOTE
--      See `LICENSE' for full license.
--
------------------------------------------------------------------------------->

# uncrlf

## Summary

A simple lexer which removes carriage return line breaks from a given file.

## License

This project's license is **GPL 2** (as of June 1991).  The whole license text
can be found in `LICENSE` in the main directory of this repository.  The brief
version is as follows:

> Copyright (C) 2022 Kevin Matthes
>
> This program is free software; you can redistribute it and/or modify
> it under the terms of the GNU General Public License as published by
> the Free Software Foundation; either version 2 of the License, or
> (at your option) any later version.
>
> This program is distributed in the hope that it will be useful,
> but WITHOUT ANY WARRANTY; without even the implied warranty of
> MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
> GNU General Public License for more details.
>
> You should have received a copy of the GNU General Public License along
> with this program; if not, write to the Free Software Foundation, Inc.,
> 51 Franklin Street, Fifth Floor, Boston, MA 02110-1301 USA.

When compiling a printable version of this documentation using Pandoc, the full
license will be attached automatically to the resulting document.  This can be
invoked by calling `repository-manual.m`.

## Software Requirements

| Requirement       | Type          | Role                                  |
|:------------------|:-------------:|:--------------------------------------|
| `docs-snippets`   | repository    | documentation constants               |
| Flex              | application   | Lex compiler                          |
| GCC               | application   | C compiler                            |
| GNU Octave        | application   | execution of the provided scripts     |
| Pandoc            | application   | compilation of repository manual      |
| `texlive-full`    | package       | compilation of repository manual      |

The compilation of such an **optional** repository manual can be invoked by just
calling one of the following lines in a terminal.

```
octave repository-manual.m
octave-cli repository-manual.m
```

Both will redirect to the same application, GNU Octave, which will then create
the manual for this repository and attach the entire license to it.  The
resulting file, `repository.pdf`, will be saved in the main directory of this
repository.

## Description

This repository provides a simple lexer in order to replace carriage return
line breaks from input source files with simple newline whitespaces.  It can be
compiled by invoking the GNU Octave script `flex-gcc.m` in a terminal:

```
octave flex-gcc.m
octave-cli flex-gcc.m
```

When optimising a source file with this lexer, the file needs to be passed from
`stdin` to the application.  The result will be written to `stdout`.

The script `install-app.m` will install the application in `~/.local/bin/`, at
option.

```
octave install-app.m
octave-cli install-app.m
```

<!----------------------------------------------------------------------------->


CHANGELOG.md
<!------------------------------------------------------------------------------
--
-- Copyright (C) 2022 Kevin Matthes
--
-- This program is free software; you can redistribute it and/or modify
-- it under the terms of the GNU General Public License as published by
-- the Free Software Foundation; either version 2 of the License, or
-- (at your option) any later version.
--
-- This program is distributed in the hope that it will be useful,
-- but WITHOUT ANY WARRANTY; without even the implied warranty of
-- MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
-- GNU General Public License for more details.
--
-- You should have received a copy of the GNU General Public License along
-- with this program; if not, write to the Free Software Foundation, Inc.,
-- 51 Franklin Street, Fifth Floor, Boston, MA 02110-1301 USA.
--
----
--
--  FILE
--      CHANGELOG.md
--
--  BRIEF
--      The development history of this project.
--
--  AUTHOR
--      Kevin Matthes
--
--  COPYRIGHT
--      (C) 2022 Kevin Matthes.
--      This file is licensed GPL 2 as of June 1991.
--
--  DATE
--      2022
--
--  NOTE
--      See `LICENSE' for full license.
--      See `README.md' for project details.
--
------------------------------------------------------------------------------->

# Changelog

## [Unreleased]

### Changed

* update submodule `routines`

### Fixed

* typo "linebreak"

## [1.0] -- 2022-04-03

### Removed

* submodule `untab`

## [0.1] -- 2022-04-03

### Added

* add GPL 2.0 license
* build instruction:  entire compilation routine
* build instruction:  installation
* build instruction:  repository manual
* create `.gitignore`
* create list of authors
* create project meta data for repository manual
* create repository README
* create this changelog
* create this repository
* create `uncrlf.l`
* submodule `docs-snippets`
* submodule `routines`
* submodule `untab`

### Deprecated

* submodule `untab`

<!----------------------------------------------------------------------------->


flex-gcc.m
%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
%%
%% Copyright (C) 2022 Kevin Matthes
%%
%% This program is free software; you can redistribute it and/or modify
%% it under the terms of the GNU General Public License as published by
%% the Free Software Foundation; either version 2 of the License, or
%% (at your option) any later version.
%%
%% This program is distributed in the hope that it will be useful,
%% but WITHOUT ANY WARRANTY; without even the implied warranty of
%% MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
%% GNU General Public License for more details.
%%
%% You should have received a copy of the GNU General Public License along
%% with this program; if not, write to the Free Software Foundation, Inc.,
%% 51 Franklin Street, Fifth Floor, Boston, MA 02110-1301 USA.
%%
%%%%
%%
%%  FILE
%%      flex-gcc.m
%%
%%  BRIEF
%%      Create a new lexer using Flex and GCC.
%%
%%  AUTHOR
%%      Kevin Matthes
%%
%%  COPYRIGHT
%%      (C) 2022 Kevin Matthes.
%%      This file is licensed GPL 2 as of June 1991.
%%
%%  DATE
%%      2022
%%
%%  NOTE
%%      See `LICENSE' for full license.
%%      See `README.md' for project details.
%%
%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%

%%%%
%%
%% Variables.
%%
%%%%

% Software.
compiler.flags  = ' -Wall -Werror -Wextra -Wno-unused-function -Wpedantic ';
compiler.in     = '*.c';
compiler.link   = ' -lfl ';
compiler.out    = 'uncrlf';
compiler.self   = ' gcc ';
compiler.call   = [compiler.self ' ' compiler.flags ' ' compiler.in];
compiler.call   = [compiler.call ' ' compiler.link ' -o ' compiler.out];

scangen.in      = 'uncrlf.l';
scangen.out     = 'lex.yy.c';
scangen.self    = ' flex ';
scangen.call    = [scangen.self ' ' scangen.in];



% Files.
files.self  = 'flex-gcc.m';



% Miscellaneous.
banner  = ['[ ' files.self ' ] '];



%%%%
%%
%% Build steps.
%%
%%%%

% Begin build instruction.
disp ([banner 'Begin build instruction.']);



% Call Flex.
disp ([banner 'Compile scanner definition with Flex ...']);

disp (scangen.call);
system (scangen.call);

disp ([banner 'Done.']);



% Call GCC.
disp ([banner 'Compile C source code with GCC ...']);

disp (compiler.call);
system (compiler.call);

disp ([banner 'Done.']);



% Clean build artifacts.
fprintf ([banner 'Remove build artifacts ... ']);

if length (glob (scangen.out));
    delete (scangen.out);
end;

disp ('Done.');



% End build instruction.
disp ([banner 'End build instruction.']);

%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%


install-app.m
%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
%%
%% Copyright (C) 2022 Kevin Matthes
%%
%% This program is free software; you can redistribute it and/or modify
%% it under the terms of the GNU General Public License as published by
%% the Free Software Foundation; either version 2 of the License, or
%% (at your option) any later version.
%%
%% This program is distributed in the hope that it will be useful,
%% but WITHOUT ANY WARRANTY; without even the implied warranty of
%% MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
%% GNU General Public License for more details.
%%
%% You should have received a copy of the GNU General Public License along
%% with this program; if not, write to the Free Software Foundation, Inc.,
%% 51 Franklin Street, Fifth Floor, Boston, MA 02110-1301 USA.
%%
%%%%
%%
%%  FILE
%%      install-app.m
%%
%%  BRIEF
%%      Install a compilation of the provided application.
%%
%%  AUTHOR
%%      Kevin Matthes
%%
%%  COPYRIGHT
%%      (C) 2022 Kevin Matthes.
%%      This file is licensed GPL 2 as of June 1991.
%%
%%  DATE
%%      2022
%%
%%  NOTE
%%      See `LICENSE' for full license.
%%      See `README.md' for project details.
%%
%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%

%%%%
%%
%% Variables.
%%
%%%%

% Software.
octave.in   = 'flex-gcc.m';
octave.out  = './uncrlf';
octave.self = ' octave ';
octave.call = [octave.self ' ' octave.in];

% Directories.
directories.dep1    = '~/.local/';
directories.target  = '~/.local/bin/';

% Files.
files.self  = 'install-app.m';

% Miscellaneous.
banner  = ['[ ' files.self ' ] '];



%%%%
%%
%% Build steps.
%%
%%%%

% Begin build instruction.
disp ([banner 'Begin build instruction.']);



% Compile application.
disp ([banner 'Compile the application to install ...']);
system (octave.call);
disp ([banner 'Done.']);



% Create directories, if required.
fprintf ([banner 'Check for ' directories.dep1 ' ... ']);
if length (glob (directories.dep1));
    disp ('Done.');
else;
    mkdir (directories.dep1);
    disp ('Created.');
end;

fprintf ([banner 'Check for ' directories.target ' ... ']);
if length (glob (directories.target));
    disp ('Done.');
else;
    mkdir (directories.target);
    disp ('Created.');
end;



% Install.
fprintf ([banner 'Install ' octave.out ' in ' directories.target ' ... ']);
movefile (octave.out, directories.target, 'f');
disp ('Done.');



% End build instruction.
disp ([banner 'End build instruction.']);

%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
